File: 0872-leaf-similar-trees/0872-leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function(root1, root2) {
    // Helper function to traverse the tree and collect the values of the leaves
    function getLeafValues(node, leafValues) {
        if (!node) {
            return;
        }
        if (!node.left && !node.right) {
            leafValues.push(node.val);
        }
        getLeafValues(node.left, leafValues);
        getLeafValues(node.right, leafValues);
    }
    
    // Collect the values of the leaves in each tree
    let leafValues1 = [];
    getLeafValues(root1, leafValues1);
    let leafValues2 = [];
    getLeafValues(root2, leafValues2);
    
    // Compare the leaf values of each tree
    if (leafValues1.length !== leafValues2.length) {
        return false;
    }
    for (let i = 0; i < leafValues1.length; i++) {
        if (leafValues1[i] !== leafValues2[i]) {
            return false;
        }
    }
    return true;
};

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        tree1 = []
        tree2 = []
        
        def preorder(node, tree):
            if node is None:
                return 
            if node.left is None and node.right is None:
                tree.append(node.val)
            preorder(node.left, tree)
            preorder(node.right, tree)
        
        preorder(root1, tree1)
        preorder(root2, tree2)
        
        if len(tree1) != len(tree2):
            return False
        
        for i in range(len(tree1)):
            if tree1[i] != tree2[i]:
                return False
        return True